### ISSUE TITLE
comparing weights

### ISSUE THREAD

**Issue by rodrigo2019** [experiencor/keras-yolo2#291]
Hello, I did some tests and I would like to share
I trained 4 models comparing the results between the same model with pre trained weights and without.
My configuration for this tests was:

Batch Size = 4
Image Size = 224x224
Dataset = VOC
Epochs = 153 (150 + 3 warmup)

## Full Yolo:

**MAP**
![image](https://user-images.githubusercontent.com/8016520/40305636-b0e574f0-5cd1-11e8-9fb7-099dd8cc7fd5.png)

**LOSS**
![image](https://user-images.githubusercontent.com/8016520/40305664-d37b3568-5cd1-11e8-8df2-352965f92075.png)

ps: After epoch 42 the model without pre trained weights started to overfit, even getting a worst result than a model with pre trained weights
**Why the model without pre trained weights got a lot of Nan values in validation?**

## Tiny yolo

**MAP:**
![image](https://user-images.githubusercontent.com/8016520/40305795-75062cd0-5cd2-11e8-804f-94188a40b1f9.png)

**LOSS**
![image](https://user-images.githubusercontent.com/8016520/40305816-87428d80-5cd2-11e8-8b47-b41a8f3c4c87.png)

In tiny yolo study case is possible to see that the model without pre trained weights get almost the same result as pre trained weights, and converged more quickly. **In this case, the pre trained weights could be wrong? Maybe was uploaded a wrong model?**

### General questions:

1. This pre trained weights was trained in the imagenet dataset (that one with 155gb and 1.2kk images)
2. The model was trained for classification?
3. What was the configuration of the training?
4. How many epochs was trained and how much hours took?
5. In this implementation, Why IOU threshold for MAP validation is 0.3 and not 0.5 as usual?

**Comment by rodrigo2019** [comment]
Hi @Axel13fr , you can check my [fork](https://github.com/rodrigo2019/keras-yolo2). I did some changes in this repository, I'am waiting @experiencor  answer my first PR to open more PR with these modifications

In my fork actually you can:

1. train with different dimensions for HxW
2. train in grayscale
3. Create a custom backend
4. mAP is avaliable on tensorboard callback for default
5. parse CSV annotations
6. If no pre trained weights are found, the model will use a fresh model instead and will not raise a error as in this repo
7. create a backup for each training (useful to compare models)
8. get inference model, excluding any dependency from the class "YOLO"

I would like to suggest all this modification for this repo, but the owner needs to answer me first

**Comment by Axel13fr** [comment]
Hi Rodrigo, 
thanks, I cherry picked your mAP Tensorboard callback. 

Regarding your questions, I believe  you have used the .h5 backend files right ? 
Have you tried using the weights from the darknet implementation to train your network ? 
This is what's done in the Jupyter Notebook using the WeightReader utils. 

As for the NaN values, initialization of the weights is very sensitive for large networks, you may want to try out something different like https://www.tensorflow.org/api_docs/python/tf/contrib/layers/xavier_initializer

**Comment by rodrigo2019** [comment]
Thanks @Axel13fr , I didn't know about the Jupyter, I will test in the next week (now I'm testing retinaNet) and post here the results

**Comment by rodrigo2019** [comment]
I did more tests, now with my particular dataset, and I got some interesting results.

In epoch 25 was the best val_loss of the training, but the mAP was 0
![bestloss](https://user-images.githubusercontent.com/8016520/40657683-18571186-631f-11e8-9f06-7efb0e7b426d.png)

In epoch 256 was the best mAP, and the val_loss was much higher than the best one.
![bestmap](https://user-images.githubusercontent.com/8016520/40657691-1e108814-631f-11e8-8d25-fbfe4fa1349f.png)
 It's look like is not a good idea to use the model with best val_loss instead best mAP
Also with batch_size = 32, I got really worst results than batch_size = 4

**Comment by Axel13fr** [comment]
Interesting. Did you do a warmup before training ? Also, have you tried to freeze some of the layers, specially during the warmup phase ? This can help to initialize the last layer of your network without causing a large gradient during training (which would affect the feature extractor badly).

**Comment by rodrigo2019** [comment]
@Axel13fr , in this test I'm using a custom backend, because I would like to run at high fps on cpu, so I customized a network based on Tiny Darknet (not Tiny Yolo), so my training is starting from scratch.
For now is too painfull to me for create pre trained weights for these custom backend, because I'm not sure if the network design will be the best choice. Also I don't known what is the best choice to generate pre trained weights for the backend, probaly I will try with autoencoders, because it looks more easier to train, but again, I don't know if is the best choice.

> Did you do a warmup before training ?

Yes, 3 epochs

> Also, have you tried to freeze some of the layers, specially during the warmup phase ? This can help to initialize the last layer of your network without causing a large gradient during training (which would affect the feature extractor badly).

It's look a very nice idea, I'm never imagined doing that, Have you ever to do that?

about the batch size, looking the discussion [here](https://github.com/fizyr/keras-retinanet/issues/325) in the retinanet, this discussion show that doesn't matter the size of yout batch, your results always will be closer of each other. It will be always truth? 
This loss function is exactly the same as the original? I didn't see the penalize factors on the original paper as here we have

In the [paper](https://arxiv.org/pdf/1506.02640.pdf) says:

> To avoid overfitting we use dropout and extensive data
> augmentation. A dropout layer with rate = .5 after the first
> connected layer prevents co-adaptation between layers [18].
> For data augmentation we introduce random scaling and
> translations of up to 20% of the original image size. We
> also randomly adjust the exposure and saturation of the image
> by up to a factor of 1.5 in the HSV color space.

I will start putting a dropout layer in the model, but first connected layer means first convolution layer?

**Comment by Axel13fr** [comment]
@rodrigo2019 this repo implements Yolo V2, which is a bit different from the original version you mention in the paper. Specifically, it uses a conv layer at the end of the network instead of fully connected layers + it uses batch normalization which is a great method to reduce over-fitting (this is why you don't see the dropout anymore in V2). 

Still though, I'm trying out a lot of stuff to train tiny yolo properly on one class, including weight freezing. It's really easy to do if you want to try, just set trainable=false for a given layer. You could do it for the couple of first conv layers as they are the least specific (do it for a complete block ie.: conv + batch norm + lrelu + MaxPool)

Example:
```

        x = Conv2D(16, (3,3), strides=(1,1), padding='same', name='conv_1', use_bias=False, trainable=trainable)(input_image)
        x = BatchNormalization(name='norm_1', trainable=trainable)(x)
```

**Comment by rodrigo2019** [comment]
in this [link](https://mlblr.com/includes/mlai/index.html#yolov2) has a good explanation about yolo

Are you successfull in your trainings?

**Comment by Axel13fr** [comment]
Thanks for the link, I'll take a look :+1: 

I've managed to train Full Yolo properly on one class on the COCO dataset but I'm still trying to have tiny yolo working properly... it seems to overfit too early but I will try to see how the MAP evolves even if the validation error goes higher.

**Comment by rodrigo2019** [comment]
by the way, on the PASCAL VOC Yolov2 got 70> on mAP evaluation, and in my tests with pre trained weights I got just 35 (with IOU threshold 0.3, with 0.5 probaly will be worst)

What was your results in the COCO dataset?

**Comment by rodrigo2019** [comment]
![image](https://user-images.githubusercontent.com/8016520/40846276-8b35f45c-658f-11e8-8f80-6e739cc81b5f.png)
 in the last 2 trains I got more than 5 hours of training resulting in mAP = 0

respective losses:
![image](https://user-images.githubusercontent.com/8016520/40846412-ee5d4742-658f-11e8-9512-7c4b800f343d.png)

**Comment by Axel13fr** [comment]
@rodrigo2019 I was training on a single class (boat) only on COCO, got 34 with Full Yolo, only 10 with Tiny, Yolo authors had 42 on 80 classes. It is worth noting though that the per class mAP can vary a lot (with 80+ for cats and as low as 22 for bottles for example on the first YOLO paper) so the average over classes can be missleading.

By the way this repo implementation never got to the published mAP because of missing training features and as well some magic training tricks of the yolo authors (and possibly some bugs left ??).

**Comment by rodrigo2019** [comment]
Did you found somewhere about the mAP on Tiny Yolo? I didn't see anywhere

Today I will convert the weights from Yolov2 (original from darknet framework) 416x416 trained on VOC and I will apply the mAP validation from this repo to compare the results. Doing this I think we can get sure if the model and the validation is correct.

**Comment by Axel13fr** [comment]
MAP for tiny Yolo is on yolo darknet website. For tiny yolo I implemented loading the darknet weights from darknet website but the model they used was for voc dataset and it had 512 filters on last conv(the one before the randomly initialized final layer) instead of 1024 as in this implementation so I had to modify it.  I will try the evaluation on voc to double check.

**Comment by rodrigo2019** [comment]
Why is necessary randomize the last conv layer? 
following the jupyter tutorial I skiped the randomize part, and I got nan values for all predictions. It is possible to use the model converting direct from darknet framework without training again?

**Comment by Axel13fr** [comment]
So you got the correct mAP for coco and voc by loading the darknet weights? 

As for the layer sizes, you can check and change them according to the cfg files from the darknet yolo site (cfg file for each network containing the topology).

**Comment by rodrigo2019** [comment]
> So you got the correct mAP for coco and voc by loading the darknet weights?

not yet, I need convert the json format, into xml or csv or create a json parser, but checking the results drawing the bounding box looks very good

> As for the layer sizes, you can check and change them according to the cfg files from the darknet yolo site (cfg file for each network containing the topology).

I knew about this but didn't know about the difference, also in the cfg file I read this:
```
object_scale=5
noobject_scale=1
class_scale=1
coord_scale=1
```
the same parameters used in this repo

if you want, you can download that conversion I did [here](https://github.com/rodrigo2019/keras-yolo2/releases/tag/trained-weights)

**Comment by Axel13fr** [comment]
I implemented the complete reading of darknet weights for full yolo and a mAP evaluation on COCO (without training) dataset gave me 45.32, a bit higher than 42.1. I think this is due in part to the obj_threshold at 0.3 instead of 0.5.
 
There is a high variance between mAP of different classes, below the details which gives a good hint of the kind of object Yolo is good at on COCO dataset (best is cats, worst is toaster):
person 0.5003
bicycle 0.3720
car 0.3260
motorcycle 0.5293
airplane 0.7025
bus 0.6725
train 0.8455
truck 0.3366
boat 0.3018
traffic light 0.2383
fire hydrant 0.7284
stop sign 0.6543
parking meter 0.3811
bench 0.3031
bird 0.3458
cat 0.8638
dog 0.7310
horse 0.6656
sheep 0.5174
cow 0.5198
elephant 0.7914
bear 0.8676
zebra 0.7483
giraffe 0.8019
backpack 0.1488
umbrella 0.3683
handbag 0.0973
tie 0.4003
suitcase 0.4296
frisbee 0.5951
skis 0.3044
snowboard 0.4185
sports ball 0.3039
kite 0.3474
baseball bat 0.3975
baseball glove 0.3562
skateboard 0.5641
surfboard 0.4926
tennis racket 0.5807
bottle 0.2178
wine glass 0.2765
cup 0.3061
fork 0.2703
knife 0.1550
spoon 0.1319
bowl 0.3513
banana 0.3014
apple 0.2374
sandwich 0.4844
orange 0.2872
broccoli 0.3941
carrot 0.2201
hot dog 0.3819
pizza 0.6292
donut 0.5248
cake 0.4822
chair 0.2927
couch 0.5671
potted plant 0.3471
bed 0.6953
dining table 0.4455
toilet 0.8429
tv 0.6970
laptop 0.6584
mouse 0.4815
remote 0.2257
keyboard 0.5966
cell phone 0.3244
microwave 0.6677
oven 0.5966
toaster 0.0525
sink 0.4845
refrigerator 0.6732
book 0.0910
clock 0.5997
vase 0.3862
scissors 0.4191
teddy bear 0.6039
hair drier 0.0580
toothbrush 0.2469

**Comment by rodrigo2019** [comment]
really nice, could you share your code?
I am interested how you read the coco annotations

**Comment by rodrigo2019** [comment]
I am trying to implement the multi-scale feature, but it looks not too easy
I didn't understand why some weights need to be initilized in this way

```python
        # initialize the weights of the detection layer
        layer = self.model.layers[-4]
        weights = layer.get_weights()

        new_kernel = np.random.normal(size=weights[0].shape)/(self.grid_h*self.grid_w)
        new_bias   = np.random.normal(size=weights[1].shape)/(self.grid_h*self.grid_w)

        layer.set_weights([new_kernel, new_bias])
```
the kernel and bias depends on grid size, and with a multi-scale the yolo model will result in a variable grid size, so I don't know how to initialize these weights.

the function loss I was expecting more work, but was quite easy to transform it to work with multi-scale, on the other hand changing the generator function is more complicated than I expected.

A fast and dirty way to implement multiscale is using the imgaug library, but in this way I think it is not safe to make zoom in in the image, because probaly in some images you may be get problem with the annotations thats goes out in the zoom area, but making just zoom out looks safe
I  already implemented it, you can check in this [branch](https://github.com/rodrigo2019/keras-yolo2/blob/ftd/augmentData/preprocessing.py#L141) if you want, but I didn't tested yet. It is very similar from the original, but now the jitter is heavier and I implemented the jitter on the Keypoints from the annotations using the same library, and not doing it manually as the original implementation

**Comment by Axel13fr** [comment]
@rodrigo2019  for the COCO dataset, the easiest is to convert it to VOC format using https://gist.github.com/chicham/6ed3842d0d2014987186#file-coco2pascal-py so that this repo can use it directly.

About the random init: this applies only to the very last layer which is supposed to be customized for your application as it depends on your settings like grid size classes etc. The rest of the architecture remains unchanged so you can do transfer learning (i.e: use pretrained weights instead of starting from zero).

In my case, as I used COCO dataset, it's the same setup as for the darknet weights so I could init this last layer with the darknet weights instead of randomly init them to do a new training.

**Comment by rodrigo2019** [comment]
but this initialize is done even in the model from frontend [here](https://github.com/experiencor/keras-yolo2/blob/master/frontend.py), when the backend weights is loaded this weights dont affect the layer which is done this init, and if you load a whole pre trained model, the pre trained weights overwrites this initilize.
why the default weights from keras is not good?

thank you about the converter coco2pascal, it will be usefull to me

**Comment by Axel13fr** [comment]
> when the backend weights is loaded this weights dont affect the layer which is done this init
That's right, this doesn't affect the front end so the frond end needs to be initialized with something appropriate to start training it from scratch and this is what this random init is doing (leaving it uninitialized will screw the training)
>  if you load a whole pre trained model, the pre trained weights overwrites this initilize
Correct, that's when you do fine tuning: you already have your full network trained and you want to adjust it a bit so you start from pretrained.

The random init is for transfer learning: you start with a backend trained for feature extraction (usually on ImageNET) and you will plug to that a front end to solve your specific problem.

**Comment by rodrigo2019** [comment]
there is some place that I can study it? Because I'm not understanding. I understand to randomize your weights if some weight is loaded in this layer, but the point is no weight is loaded in this specific layer 
previously

**Comment by Axel13fr** [comment]
That's for the case where no weights are loaded : then what are the values of uninitialized variables of this layer ?  
0 would be a terrible case. High values could be as well. So better not leave that undefined: simply use an init method, random could be one of them. 

There are many sources about this subject which you may want a look at. For example https://stats.stackexchange.com/questions/200513/how-to-initialize-the-elements-of-the-filter-matrix?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa

**Comment by rodrigo2019** [comment]
keras uses for default a initilizer called "glorot_uniform", but for this layer the kernel initializer is "lecun_normal".
```python
        # make the object detection layer
        output = Conv2D(self.nb_box * (4 + 1 + self.nb_class), 
                        (1,1), strides=(1,1), 
                        padding='same', 
                        name='DetectionLayer', 
                        kernel_initializer='lecun_normal')(features)
```

thank you

**Comment by Axel13fr** [comment]
You're absolutely right, I overlooked that. Seems like this random uniform call is useless after the kernel initializer from keras!

**Comment by rodrigo2019** [comment]
maybe @experiencor  was using lecun_normal initializer before implementing this new initializer which maybe has in the original implementation on darknet framework

**Comment by lichacha** [comment]
if i want change the net of full yolo, but i don't have the corresponding backend.h5 .what can i do for it? 
Read through the code,I find that users can't seem to modify the network of full yolo or tiny yolo. aaaaaaa What should I do?

**Comment by bkanaki** [comment]
@rodrigo2019 @Axel13fr Do you guys have the weights converted for the Tiny-YOLO model from darknet to this framework?

**Comment by rodrigo2019** [comment]
@bkanaki just for full yolo on coco, but you can create this model using the tools provided in this repo.
If you do, please share with us, I'am uploading the models [here](https://github.com/rodrigo2019/keras-yolo2/releases)

**Comment by bkanaki** [comment]
I somewhere read on the issues that it is difficult to reproduce results for Tiny-Yolo model. The darknet website reports it to be 23% on the COCO dataset. I will see if I can get good results and then I can post whenever I get time.

**Comment by bkanaki** [comment]
Hi @Axel13fr @rodrigo2019 
I have a one more question for you guys:

When you tried training from scratch, i.e., without using the pre-trained weights, did you ever see the loss going to `nan`? For me, from very beginning it is `nan` and doesn't change if I train only on the images with person from coco dataset.

@rodrigo2019 , you had a question if the loss implemented here is the exact loss as mentioned in the paper. I think this loss doesn't take the square root of the width and height predictions but the original paper mentions it. 

Even the snippet in the Step-By-Step notebook, the author of this repo (@experiencor) mentions square root, but I don't see it in the actual definition. Any reason why? Or is it overlooked?

**Comment by rodrigo2019** [comment]
Yes I got nan values in some training, even training from scratch.
Did you tried putting the square root?

**Comment by bkanaki** [comment]
Yes, I put square root. When debugging the loss function, I noticed that my loss was blowing up because of the `loss_wh`. That is when I noticed that `sqrt` is missing. This brought down the loss quickly after warmup (I am aware of the +10 added to loss during warmup) than compared to without square root. My mAP also increased from 25% to 68% for my dataset.

However, the `nan` might also have to do with the different anchors, as someone has posted in the issues.

I still haven't attempted training from scratch, but I will as I have to do something very similar to what you did, i.e., reducing network size further.

**Comment by rodrigo2019** [comment]
wow really interesting, could you share the code? I would like to try

**Comment by bkanaki** [comment]
Just add:

```
true_box_wh = tf.sqrt(true_box_wh)
pred_box_wh = tf.sqrt(pred_box_wh)
```

before [line 215](https://github.com/experiencor/keras-yolo2/blob/4e8c85ce02435f136d4f4cfe930b4ccb759fbaf8/frontend.py#L215)

**Comment by Axel13fr** [comment]
@bkanaki I did write a function to read the darknet weights for Tiny Yolo, it's an adaptation of the existing one which reads Full Yolo darknet weights: 

`    def read_weights(self):
        weight_reader = WeightReader(TINY_YOLO_OFFICIAL_WEIGHTS_PATH)
        weight_reader.reset()
        nb_conv = 9
        last_layer_with_batch_norm = 7
        # Only reads weights for the first 8 layers. The last one is init randomly
        # outside the constructor. This lets the user choose the side of the output :
        # number of boxes, classes
        for i in range(1, nb_conv+1): #1:8
            # For layer 1 to 7 : read batchnorm
            if i <= last_layer_with_batch_norm:
                norm_layer = self.model.get_layer("norm_" + str(i))

                size = np.prod(norm_layer.get_weights()[0].shape)

                beta  = weight_reader.read_bytes(size)
                gamma = weight_reader.read_bytes(size)
                mean  = weight_reader.read_bytes(size)
                var   = weight_reader.read_bytes(size)

                weights = norm_layer.set_weights([gamma, beta, mean, var])

            conv_layer = self.model.get_layer("conv_" + str(i))
            if len(conv_layer.get_weights()) > 1:
                # Read bias & weights
                bias = weight_reader.read_bytes(np.prod(conv_layer.get_weights()[1].shape))
                kernel = weight_reader.read_bytes(np.prod(conv_layer.get_weights()[0].shape))
                kernel = kernel.reshape(list(reversed(conv_layer.get_weights()[0].shape)))
                kernel = kernel.transpose([2, 3, 1, 0])
                conv_layer.set_weights([kernel, bias])
            else:
                # Read only weights
                kernel = weight_reader.read_bytes(np.prod(conv_layer.get_weights()[0].shape))
                kernel = kernel.reshape(list(reversed(conv_layer.get_weights()[0].shape)))
                kernel = kernel.transpose([2, 3, 1, 0])
                conv_layer.set_weights([kernel])
`

**Comment by jzx-gooner** [comment]
@rodrigo2019 Thanks for your share !
I want to train a yolo model with only one class.
I have trained the model based on the Full Yolo and Tiny Yolo weights 
**The result :**
- Full Yolo: It is good ,however the model size is too big and run slow in cpu.

- Tiny Yolo:It is bad.There are lots of  Misjudgement.

I have three quenstions:

- How to improve the Tiny Yolo's accuracy.

- Could I  design a small network replace the Yolo cnn network based on my dataset

- As it was stated,Can small neural networks achieve real-time performance on CPU?
Thank you very much！

**Comment by rodrigo2019** [comment]
@jzx-gooner,
yes you can archive real-time performance on CPU, I have a network for car detection that runs over 15fps on hardware like raspberry pi 3, on a notebook core i7 it runs over 150~200fps.

To do it, you must design a small backend with a small input size, like 50x50. But this network will not be good prediction small objects in the image

**Comment by jzx-gooner** [comment]
> @jzx-gooner,
> yes you can achive real-time performance on CPU, I have a network for car detection that runs over 15fps on hardware like raspberry pi 3, on a notebook core i7 it runs over 150~200fps.
> 
> To do it, you must design a small backend with a small input size, like 50x50. But this network will not be good prediction small objects in the image

Thank your for your kindly reply. I used your repository and use the supertinyyolo which your design,i train the model from  scratch。The model is little and fast，however the accuracy is low and three are lots of  wrong judgement。

**Comment by rodrigo2019** [comment]
I don't know how hard is your dataset, in my data all samples are very similar, although I got some false positives the network is working fine. Try to add some more filters in each layer

**Comment by martinbel** [comment]
@rodrigo2019 @Axel13fr @experiencor I've extended rodrigo's repo to support conversion of models to the NCS movidius stick. Take a look if you have some time. Thanks for sharing your work!

**Comment by rodrigo2019** [comment]
@martinbel really nice, I have a few questions about your work.
1 - How many times the movidius stick runs faster than raspberry pi ? Are you using the pi3?
2 - How good is the accuracy of your model? (mAP)

ps: I think you can improve your network, I would try using a smaller image like 128x128, decrease the amount of filters in the firsts layers, and increase the filters in the last layers.

**Comment by rodrigo2019** [comment]
> I think there might be a bug in the decode_netout, where NMS is done but I'm not sure. I'm refactoring it a bit, trying to avoid making massive changes.

probaly, few weeks agor I found a bug in the NMS code and fixed it, maybe there are more bugs

> I wasn't able to make tensorboard work. Any ideas what could be the problem?

Are you using the back option in the config file? tensorboard is working fine to me.

> So far I haven't been able to get "great" results by even training a detector with one class. I'll give your network suggestion a try.

I was able to get mAP higher than 70% for a car detection, but I used a very specific dataset with specific position for all samples, the coco dataset is much harder than my dataset.

**Comment by martinbel** [comment]
I've been trainign a model for 1 class and I was able to get 0.85 mAP. 
It seems to be generalizing well also. 
Here is the setup: 
- Using rodrigo's fork. I've made some changes to nms, the rest is the same.
- Backend: rodrigo's SuperTinyYolo backend. 
- 1 anchor box: [1,1], this is for a fairly small object.
- Input size: 256, 256, 1 
- I've included negative instances also, this helped reduce the false positives. For this I'm just including images with other objects and no annotation. An empty .xml file
- I have around 350 labeled images, so I upsample them to ~1000 and use 1000 negative instances.
- To train I use 1e-3, then 1e-4

**Comment by rodrigo2019** [comment]
Did you used just 1 anchor box? or this is a specific one for small objects and you used more than a single one?

**Comment by martinbel** [comment]
Just one anchor box. I think this makes sense as it's just one object and the Detection Layer has 16x16.
I've also made some changes to the NMS algo, I'll commit this to my repo.

**Comment by rodrigo2019** [comment]
Thank you for sharing, I will try it in my models as well

**Comment by jzx-gooner** [comment]
> ss loss than the bounding

Thank you for sharing.

**Comment by martinbel** [comment]
@jzx-gooner I haven't used the csv parser but it should be the same. So far, I was able to train models with one class only. I still think the original and rodrigo's NMS are off. It's enough to see one example of the predictions to see it's not really working. With the changes I made, NMS works much much better and the model learns in less iterations.

**Comment by jzx-gooner** [comment]
> > @jzx-gooner I haven't used the csv parser but it should be the same. So far, I was able to train models with one class only. I still think the original and rodrigo's NMS are off. It's enough to see one example of the predictions to see it's not really working. With the changes I made, NMS works much much better and the model learns in less iterations.

Thank you for your help. I can see the code in csv mode in the line [91](https://github.com/rodrigo2019/keras-yolo2/blob/11f3e325cabb0007bd156f005f79611a5264ff6a/preprocessing.py#L91). It deals with the images with no annotation,However,i could not find the same code in xml mode,maybe it not work correctly?

**Comment by clw5180** [comment]
Hello, rodrigo2019. I really learned a lot from your reply on keras-yolo2, and I have some doubts for the NMS. You said 
> probaly, few weeks agor I found a bug in the NMS code and fixed it, maybe there are more bugs
and I compare your code with the author @experiencor 
```
                    if bbox_iou(boxes[index_i], boxes[index_j]) >= nms_threshold:
                        boxes[index_**_i_**].classes[c] = 0
```
You wrote 'boxes[index_**i**]' in the second line, however the author wrote boxes[index_**j**]. I think the author is right, because boxes[index_i] 's score is higher than boxes[index_j] (because i always less than j, and has code 'reversed(np.argsort')) , so if bbox_iou >= nms_threshold, the boxes[index_j] should be removed.

**Comment by rodrigo2019** [comment]
@clw5180 I understand your point of view, but using index_j you will get results like this:
Here you can see clearly that the IOU from theses box are higher than 0.5, and just one of these box should be selected, but instead it keeps the both boxes
![id5423-anj8921-1original](https://user-images.githubusercontent.com/8016520/50519358-a8d42a80-0aa1-11e9-94b6-a8c63e5df077.jpg)
 When the correct results should be (using index_i):
![id5423-anj8921-1original](https://user-images.githubusercontent.com/8016520/50519368-b689b000-0aa1-11e9-99a8-3ace183d37a4.jpg)

I Understand that using index_i we are losing the best score, but in this way the boxes will be selected in a better way. I think that NMS code should be improved, I tried to use the NMS function from opencv library, but I got worst results, I don't know what kind of mistake I made

**Comment by clw5180** [comment]
> I made two modifications in the loss function, but I didn't checked the recall values after the modifications, I will try to figure it out.
> Are you using both repos? Could you share your results? which repository give you a better results?
> Using my fork I always get better mAP results, but sometimes looking in the results I Think the original function wrote by @experiencor is better than my function with the modifications.
> 
> Also I tried to train with the original Darknet using the same backend network design, and using this version of yolo I got better results than using the original darknet

@rodrigo2019 Hi rodrigo, I'm just a noob and I don't know how to use the tensor board. I just make the debug item 'true' to check whether the Average_Recall is going up. Now I take part in a competition about head-counting, the training dataset has 300 pictures, and each picture has lots of people(i.e. lots of head, ~ 100+). However, for just a few epoches, the val_loss stop to improve so there's no .h5 model saved, but the Average Recall is still very low and the prediction result is bad. Yesterday I learn that Yolo is not suitable for intensive object detection, should I try to use faster R-CNN? Do you have any good ideas or models? Thanks a lot.

**Comment by rodrigo2019** [comment]
maybe retinanet can be better than yolo. But for the yolo you should try to use a bigger output grid, you can do this using a bigger resolution or changing the backend design, using less max poolings layers. I think that less max pooling layers should be a better option

**Comment by martinbel** [comment]
@clw5180 What is  intensive object detection? BTW I've found a project that is interesting perhaps to combine to rodrigo's version, [dourflow](https://github.com/ksanjeevan/dourflow/).
It doesn't have as many features but I think the loss and NMS are correct. 
I've tried fine tunning and it worked quite well, much better than this project. But I didn't try training from scratch.

**Comment by rodrigo2019** [comment]
@martinbel I think he is saying about images with high number os objects, like peoples in a stadium.

About the dourflow, it looks nice, I just changed the loss function from this work to the douflow, and now I'am waiting for the results. But I'am expecting bad results, because I think we need make the bacth generator compatible as well, also I saw that repo just work with square images, so it would be nice if we could modify it to work as my repo, I think that is a nice feature to work with non square images.
I'am not happy with the actual loss function, because I never got the same results in VOC dataset as the original darknet. But I have a curious issue, using this repository and a specific dataset I got much better result using the @experiencor version than the original darknet, using this repo I got mAP over 90%, and using darknet I got no detections at all. In order to make the darknet work, I need to use a much bigger backend.

Also it would be nice to change the mAP callback into a mAP metric in the keras, It is too slow to apply the mAP callback every epoch,

**Comment by rodrigo2019** [comment]
@martinbel check this my [branch](https://github.com/rodrigo2019/keras-yolo2/tree/ftd/refactoredLossFunction), I changed the losses and it worked. Could you tell if the results is better than the older loss?

**Comment by clw5180** [comment]
> @martinbel check this my [branch](https://github.com/rodrigo2019/keras-yolo2/tree/ftd/refactoredLossFunction), I changed the losses and it worked. Could you tell if the results is better than the older loss?

Hi, @rodrigo2019 ，it's been a long time. I've learned how to use the tensorboard, and recently I 
 read the paper about RetinaNet that you recommended before, I think it's very suitable for the dense object detection, so I prepare to try the keras version implement on github. Before training I want to ask you a question: the total dataset has about 9000+ pictures with about 400K objects (people's head), and this large dataset comes from 5 different small datasets, some have few pictures(~ 300) but each picture have lots of objects(~ 100+), some have lots of pictures(~ 5000) but each picture have few objects(~ 5 or less), the pictures of 5 datasets come from different places such as mall, street, park etc.  , and for different dataset the sizes of pictures are also different, (some 1024*768, some 416*288). So should I train them together to get one model? (maybe it's too big and hard to train a good model?)  Or training 5 datasets separately and finally get five models? Sorry to disturb you.

**Comment by martinbel** [comment]
@rodrigo2019 great! I'll give it a try!

**Comment by martinbel** [comment]
@rodrigo2019 I tried the same model I had previously used but the loss is taking too long to decrease. After 35 epochs I don't see a reasonable result yet. 
I tried the dourflow project with pre-trained weights and fine tunning full yolo worked well but I didn't get to try training from scratch.

**Comment by rodrigo2019** [comment]
@martinbel I tried with two datasets, a small one worked good from scratch, the bigger one (PASCAL VOC) didn't worked from scracth, I trained over 24hours and get no results.

I also checked the batch generator from both repos, and I could say that they are pretty much the same code. It is really similar.

Could you explain how did you use the pre trained weights and how do you do the fine tuning? I would like to reply your steps in this repo

**Comment by rodrigo2019** [comment]
> So should I train them together to get one model? (maybe it's too big and hard to train a good model?) Or training 5 datasets separately and finally get five models? 

@clw5180 it is a hard question to know the answer, the best way is to try the both options, but probably using one model for each dataset you will get a better results, becausing you are going to fit each model in a specific conditions, but in another hand a single model can fit as well and cost 5x times less.
Try to train using a single model, if you get bad results try your second option, if both options fails, try to modify the backend from your model

**Comment by martinbel** [comment]
> @martinbel I tried with two datasets, a small one worked good from scratch, the bigger one (PASCAL VOC) didn't worked from scracth, I trained over 24hours and get no results.
> 
> I also checked the batch generator from both repos, and I could say that they are pretty much the same code. It is really similar.
> 
> Could you explain how did you use the pre trained weights and how do you do the fine tuning? I would like to reply your steps in this repo

I followed the repos readme, in the end it shows how to train. First by just making predictions with the model provided, then you can do fine tuning. 

Create a labels and anchors file in models. See the coco example.

Then split the data, place images and xml's in a separate folder.
`python3 split_dataset.py -p 0.8 --in_ann labels/ --in_img images/ --output split/`

As I was detecting 2 small objects (guns and headwear) I used 2 anchors I choose manually. 
Then you can just fine tune the full yolo model after adapting the config.
`python3 dourflow.py train -m models/coco_model.h5 -c confs/config_custom.json`

Just to clarify the fine tunning results are ok, I think better than the keras-yolo2 repo but in my case the results weren't perfect, perhaps because the objects were hard to predict.

**Comment by rodrigo2019** [comment]
the blue curve is using keras-yolov2 repo with dourflow loss function
the red curve is using keras-yolov2 from my fork

Same dataset and same backend for both. Started from scratch
![image](https://user-images.githubusercontent.com/8016520/50608889-183c8680-0eb5-11e9-9aca-ae71dce93bc7.png)

**Comment by martinbel** [comment]
@rodrigo2019 That's great! Can you give more details about the backend? What learning rate did you use to train the model?

**Comment by martinbel** [comment]
@rodrigo2019 I've been testing the code and it's working better for my use case also with the latest changes. I still didn't find the NMS code to be working well in my case so I made a few changes that are producing a better result. 
Here is a [gist](https://gist.github.com/martinbel/a60d4d3ce579c9954bc0b40d667b0e4b) with the functions that changed, perhaps you can test it out also.

**Comment by hieifn** [comment]
@rodrigo2019  Hi Rodrigo. First of all, thanks for the contributions, i'm learning a lot with you here. Well i'm just starting on YOLO model, and i'm trying to use your implementation, but i can't seem to make the tensorboard work properly. I can only see the YOLO architecture on tensorboard since i'm poniting logdir to "./logs/exemples". I'm trying to train a face detection model on Yolo using widerface dataset. Everything works, but the tensorboard.
This is my confg file:
![image](https://user-images.githubusercontent.com/17056166/53400307-54a7d280-398c-11e9-9566-6c098ed0d516.png)

If you , or anyone, could give me a few tips i would appreciate.
Thanks in advance

**Comment by rodrigo2019** [comment]
@hieifn, you are saying that the tensorboard callback creates the event file, but when you try to read this event file with tensorboard you can only see the graph model, but you can't see the metric, like mAP?

**Comment by hieifn** [comment]
@rodrigo2019 Yes! I'm probably doing something wrong, but i don't know what.

**Comment by rodrigo2019** [comment]
@hieifn  I tried to reproduce you error, but I couldn't.
also I invite you to try my new [branch](https://github.com/rodrigo2019/keras_yolo2/tree/ftd/refactoredLossFunction) which I refactored the loss function, in my testes I got better mAP results, but I need more tests before merge into the master branch.

**Comment by hieifn** [comment]
@rodrigo2019 So, i've figured out what was happening. i was tranning on my CPU (i5-4650k) and it was taking too long to process one epoch, but since i was with debug set on true i got confused. I'm now renting a AWS instance to get around this problem. By the way, what is your PC configuration to run these models ?

**Comment by rodrigo2019** [comment]
@hieifn i7-2600k, 8gb ddr3 and gtx1070ti (8gb)

**Comment by hieifn** [comment]
@rodrigo2019 Rodrigo, once again. Thank you so much for the the help.
